tower-buffer/src/lib.rs

```diff
@@ -42,18 +48,87 @@ pub struct ResponseFuture<T> {
 /// Errors produced by `Buffer`.
 #[derive(Debug)]
 pub enum Error<T> {
+    /// The `Service` call errored.
     Inner(T),
+    /// The underlying `Service` failed.
     Closed,
 }
 
+/// An adapter that exposes the associated types of a `DirectService` through `Service`.
+/// This type does *not* let you pretend that a `DirectService` is a `Service`; that would be
+/// incorrect, as the caller would then not call `poll_service` and `poll_close` as necessary on
+/// the underlying `DirectService`. Instead, it merely provides a type-level adapter which allows
+/// types that are generic over `T: Service`, but only need access to associated types of `T`, to
+/// also take a `DirectService` ([`Buffer`] is an example of such a type).
+pub struct DirectServiceRef<T> {
+    _marker: PhantomData<T>,
+}
+
+impl<T, Request> Service<Request> for DirectServiceRef<T>
+where
+    T: DirectService<Request>,
+{
+    type Response = T::Response;
+    type Error = T::Error;
+    type Future = T::Future;
+
+    fn poll_ready(&mut self) -> Result<Async<()>, Self::Error> {
+        unreachable!("tried to poll a DirectService through a marker reference")
+    }
+
+    fn call(&mut self, _: Request) -> Self::Future {
+        unreachable!("tried to call a DirectService through a marker reference")
+    }
+}
+
+/// A wrapper that exposes a `Service` (which does not need to be driven) as a `DirectService` so
+/// that a construct that is *able* to take a `DirectService` can also take instances of
+/// `Service`.
+pub struct DirectedService<T, Request>
+where
+    T: Service<Request>,
+{
+    inner: T,
+    _marker: PhantomData<Request>,
+}
+
+impl<T, Request> DirectService<Request> for DirectedService<T, Request>
+where
+    T: Service<Request>,
+{
+    type Response = T::Response;
+    type Error = T::Error;
+    type Future = T::Future;
+
+    fn poll_ready(&mut self) -> Result<Async<()>, Self::Error> {
```

Comment:
Just a small thing, but noticed this was `Result<..>` instead of `Poll<..>`, any particular reason?